### PR TITLE
KeyPathIterable

### PR DESCRIPTION
**Pull request by Azoy**
Sorry this took a lot longer than expected, but hopefully we can start pushing this through! Please make changes you feel are necessary. Some of the metadata comments aren't useful, maybe we want more tests? Please let me know.

Edit: The readme that I updated is just a short snippet, we can update it more.

cc: @rxwei @dan-zheng

### CODE REVIEW

**Comment by natecook1000** [comment]
This is exciting! 🎉 As you expand on the proposal, I have a few thoughts/questions on the proposed APIs. We don't necessarily need to discuss these here — the forums or a follow-up proposal would make sense, as well.

- It will be important to explain the difference between getting all the key paths for a type and for a value — I'm not sure those operations should have the same names. (`MemoryLayout`'s APIs could be a guide here —something like  `Reflection<Type>.allKeyPaths()` vs `Reflection.allKeyPaths(for: value)`.)
- Should array elements be treated as "named", given that we currently have the duality of keyed / unkeyed collections in both 
`Codable` and `Mirror`? What motivates the current design? What are non-named elements in this case?
- Should the custom key-path iterable functionality be provided for collections more generally than array & dictionary?
- Have you explored lazily generating the resulting collection of key paths, instead of returning an array? Getting the key paths for a collection this way will be an O(n) operation.
- Given that you can pass any type to `Reflection.allKeyPaths(for:)`, `KeyPathIterable` seems like a misnomer. Should that protocol have the `Custom` prefix, like `CustomStringConvertible` and `CustomReflectable`?
- From a prioritization standpoint, why should `KeyPathIterable` come before the tools to access type metadata that you've defined internally?

File: Sources/SE0000_KeyPathReflection/KeyPathInitialization.swift [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    
+    // Return an offseted base after the kvc string pointer.
+    let newBase = base + MemoryLayout<Int>.size
+    let newBytes = bytes - MemoryLayout<Int>.size
+    
+    body(UnsafeMutableRawBufferPointer(start: newBase, count: newBytes))
+    
+    return unsafeBitCast(object, to: self)
+  }
+}
```
Comment (jckarter) [inline Sources/SE0000_KeyPathReflection/KeyPathInitialization.swift:62]: The layout of AnyKeyPath is not ABI-stable. This is not guaranteed to work.

Comment (Azoy) [reply Sources/SE0000_KeyPathReflection/KeyPathInitialization.swift:62]: The idea behind this is simply to have something separate and usable before baking in the standard library. This already exists there (so if it goes in the stdlib, this goes away entirely), so before committing to the Swift project I had to reimplement this to have something usable.

**Comment by Azoy** [comment]
@natecook1000 Hi Nate, I believe @rxwei and/or @dan-zheng said they were going to start discussion back up on the forums. I would love to continue discussing some of the API decisions we made over there once they're ready to start discussion back up.

File: Sources/SE0000_KeyPathReflection/Reflection.swift [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    }
+    
+    return result
+  }
+  
+  /// Returns the collection of all key paths of this value.
+  ///
+  /// - Parameter value: A value of any type to return the stored key paths of.
+  /// - Returns: An array of partial key paths for this value.
+  public static func allKeyPaths<T>(for value: T) -> [PartialKeyPath<T>] {
```
Comment (bartchr808) [inline Sources/SE0000_KeyPathReflection/Reflection.swift:38]: @Azoy what do you think about adding versions where you can directly specify the destination type of the key path you want like:
```swift
static func allKeyPaths<Source, Destination>(
    from value: Source, to destinationType: Destination.Type
) -> [KeyPath<Source, Destination>] {
    allKeyPaths(from: value).compactMap { kp in
        kp as? KeyPath<Source, Destination>
    }
}
```
And also adding the writable key path version:
```swift
static func allWritableKeyPaths<Source, Destination>(
    from value: Source, to destinationType: Destination.Type
) -> [WritableKeyPath<Source, Destination>] {
    allKeyPaths(from: value).compactMap { kp in
        kp as? WritableKeyPath<Source, Destination>
    }
}
```
And finally the recursive versions:
```swift
static func recursivelyAllKeyPaths<Source>(
    for value: Source
) -> [PartialKeyPath<Source>] {
    var result = allKeyPaths(for: value)
    var workList = result
    while let kp = workList.popLast() {
        func helper<T>(_ x: T) {
            let keyPaths = allKeyPaths(for: x).map { kp.appending(path: $0)! }
            result.append(contentsOf: keyPaths)
            workList.append(contentsOf: keyPaths)
        }
        _openExistential(value[keyPath: kp], do: helper)
    }
    return result
}

static func recursivelyAllKeyPaths<Source, Destination>(
    from value: Source, to destinationType: Destination.Type
) -> [KeyPath<Source, Destination>] {
    recursivelyAllKeyPaths(for: value).compactMap { kp in
        kp as? KeyPath<Source, Destination>
    }
}

static func recursivelyAllWritableKeyPaths<Source, Destination>(
    from value: Source, to destinationType: Destination.Type
) -> [WritableKeyPath<Source, Destination>] {
    recursivelyAllKeyPaths(for: value).compactMap { kp in
        kp as? WritableKeyPath<Source, Destination>
    }
}
```

**Comment by rxwei** [comment]
After some discussion with @natecook1000 we'll merge this into the `reflection` branch and iterate further with new PRs.

**Pull request merged**